### PR TITLE
CACTUS-964 DataGrid Columns

### PR DESCRIPTION
**Pull request by wilysword**
https://repayonline.atlassian.net/browse/CACTUS-964

Includes a couple of unrelated fixes. The `ScreenSizeProvider` fix  was because I was trying to debug something and got annoyed that there was an extra render cluttering up my debug info. The Breadcrumb fix popped up when I looked at https://github.com/repaygithub/cactus/pull/775 and saw random unrelated screenshots had changed (still not sure why they picked now to change, maybe somehow because of the `ScreenSizeProvider` fix?).

The main change is that I combined Column and DataColumn into a single component. To make that a little easier I moved most of the column logic into a separate file, and then rewrote the rendering logic in DataGridTable. I doubt we'd ever see enough columns where the performance difference between a Map and an Array would be significant, so I got rid of the Map and just kept the whole thing as a sorted array.

With the table rendering, that was really too complex to keep inline the way it was. Although it's a bit simpler now that we have only one column type, I still thought it best to split the header & body cell logic into separate functions, and I tried to simplify the sorting logic a bit.

Other changes:
- Added `order` prop, defaults to column render order.
- Added `headerProps` prop, to set props on just the header cells.
- Fixed `isCardView` variable to take `variant` into account.
- Changed DataGridType, using the actual component types instead of generic ComponentType.
- Removed/simplified some styles:
  - `IconWrapper` only added `flex-shrink: 1`, which is a default style.
  - `TextWrapper` added `flex-grow: 1`, but since the flex container itself is doesn't grow, that does nothing as far as I could tell (even on IE).
  - `.flip-chevron` class replaced by simply rendering `NavigationChevronUp` instead of transforming the Down icon.
  - `.has-content` class replaced using the `:empty` pseudo-class.
  - Defined the TopSection's `spacing` as CSS, instead of cloning & adding a style prop like it was before.

I really wanted to get rid of that `topSectionRendered` weirdness, but I
felt that was just too far out of scope (as opposed to all the other
changes I made to `TopSection` ;).

### CODE REVIEW

**Comment by wilysword** [comment]
Not sure what's up with the snapshot change. The div is empty, and it has no margin, and I can't find anything _else_ with a margin, but somehow just having an empty div there adds some whitespace.

File: modules/cactus-web/src/ScreenSizeProvider/ScreenSizeProvider.tsx [hunk 1]
```diff
@@ -54,42 +51,45 @@ export const ScreenSizeContext = React.createContext<ScreenSize>(SIZES[DEFAULT_S
...
-      const media = window.matchMedia(theme.mediaQueries[mq].replace(/^@media /, ''))
-      queries[mq] = media
-      media.addListener(listener)
+  const [currentSize, setSize] = React.useState<Size>(() => {
+    queries = createQueries(theme)
+    return getMatchedSize(queries)
+  })
+
+  React.useEffect(() => {
+    const q: QueryType = queries || createQueries(theme)
```
[the remark this answers is not in the document]
Comment (wilysword) [reply modules/cactus-web/src/ScreenSizeProvider/ScreenSizeProvider.tsx:80]: `queries` is only assigned on the first render (when the state is being initialized), but if the theme changes then the effect hook could run again on subsequent renders. It's true that you wouldn't expect that to happen in normal usage, but it _can_ happen, and it _does_ happen in Storybook when you change the theme.

File: modules/cactus-web/src/ScreenSizeProvider/ScreenSizeProvider.tsx [hunk 2]
```diff
@@ -54,42 +51,45 @@ export const ScreenSizeContext = React.createContext<ScreenSize>(SIZES[DEFAULT_S
...
+    const listener = () => setSize(getMatchedSize(q))
+    for (const key of Object.keys(q)) {
+      q[key as Size].addListener(listener)
```
[the remark this answers is not in the document]
Comment (wilysword) [reply modules/cactus-web/src/ScreenSizeProvider/ScreenSizeProvider.tsx:83]: Didn't even notice that, I'll fix it to only use `addListener` if `addEventListener` doesn't exist.

[the remark this answers is not in the document]
Comment (wilysword) [reply modules/cactus-web/src/ScreenSizeProvider/ScreenSizeProvider.tsx:83]: Actually, I'm going to make that a separate ticket. This fix already shouldn't have been part of this ticket, and it's working for the foreseeable future, so it's not urgent.

**Review by daniloPenaR — APPROVED** [review]
Changes look pretty good to me in general. Thanks for simplifying the logic of this component. I agree with the fact that the deprecated `addListener` will not affect this in the near future so I'll approve this.

**Review by Dhalton — APPROVED** [review]
LGTM

**Pull request merged**